dotexporter.py

```diff
@@ -87,23 +96,116 @@ def do_GET(self):
+               'name': 'dot_chain_stall',
+               'prop': { 'block': 'best'},
+               'value': 0
+             })
+
 
         try:
           chain_getFinalizedHead   = self.query("chain_getFinalizedHead")
           chain_FinalizedHeadBlock = self.query("chain_getBlock", [chain_getFinalizedHead])
-          m.append({
-            'name': 'dot_chain_block_number',
-            'prop': { 'block': 'finalized' },
-            'value': int(chain_FinalizedHeadBlock['block']['header']['number'], 16)
-          })
+          self.current_finalized = int(chain_FinalizedHeadBlock['block']['header']['number'], 16)
+          final_drift=int(datetime.utcnow().timestamp()- self.previous['finalized_epoch'])
+
+          if self.current_finalized > self.previous['finalized']:
+            self.previous['finalized'] = self.current_finalized
+
+            if(datetime.utcnow().timestamp() >= self.previous['finalized_epoch']):
+                self.previous['finalized_epoch'] = int(datetime.utcnow().timestamp())
+                if DEBUG:
+                  print("finalized: datetime.utcnow().timestamp()" + str(datetime.utcnow().timestamp()))
+                m.append({
+                  'name': 'dot_chain_block_number',
+                  'prop': { 'block': 'finalized' },
+                  'value': self.current_finalized
+                }),
+                m.append({
+               'name': 'dot_chain_stall',
+               'prop': { 'block': 'final'},
+               'value': 0
+             })
+
+            if DEBUG:
+              print("current_final: " + str(self.current_finalized))
+              print("previous_final: " + str(  self.previous['finalized']))
+              print("previous_final_epoch: " + str(  self.previous['finalized_epoch']))
+              print("finalized seconds difference = " + str(int(datetime.utcnow().timestamp())- self.previous['finalized_epoch']))
+
+            if((datetime.utcnow().timestamp() - self.previous['finalized_epoch']) > TIMEOUT):
+             if DEBUG:
+                print("self.current_finalized: " + str(self.current_finalized))
+                print("selfprevious.['finalized_epoch']: " + str(self.previous['finalized_epoch']))
+             self.previous['finalized_epoch'] = int(datetime.utcnow().timestamp())
+             m.append({
+               'name': 'dot_chain_stall',
+               'prop': { 'block': 'final'},
+               'value': 0
+             })
+
+          else:
+           if((datetime.utcnow().timestamp() - self.previous['finalized_epoch']) > TIMEOUT):
+             m.append({
+              'name': 'dot_chain_stall',
+              'prop': { 'block': 'final'},
+              'value': 1
+              })
+             if DEBUG:
+               print("stall detected for final block")
+           else:
+             m.append({
+               'name': 'dot_chain_stall',
+               'prop': { 'block': 'final'},
+               'value': 0
+              }),
+
```

Comment:
for all the appends of the `dot_chain_stall` metric I suggest to put it down to the list of metrics and use a variable to transport the value there. otherwise you're repeating the same information and the code is less readable imo.